File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(const string& s) {
        stack<char> stk;
        for (char c : s) {
            switch (c) {
                case '(':
                case '{':
                case '[':
                    stk.push(c);
                    break;
                case ')':
                case '}':
                case ']':
                    if (stk.empty() || stk.top() != matchingOpeningBracket(c)) {
                        return false;
                    }
                    stk.pop();
                    break;
            }
        }
        return stk.empty();
    }
    
private:
    char matchingOpeningBracket(char closingBracket) {
        if (closingBracket == ')') {
            return '(';
        } else if (closingBracket == '}') {
            return '{';
        } else {
            return '[';
        }
    }
};

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        stack = []
        for i in s:
            if(i == '(' or i == '{' or i == '['):
                stack.append(i)
            else:
                if(len(stack) == 0):
                    return False
                if(i == ')' and stack[len(stack)-1] != '('):
                    return False
                if(i == '}' and stack[len(stack)-1] != '{'):
                    return False
                if(i == ']' and stack[len(stack)-1] != '['):
                        return False
                stack.pop()
        return (len(stack) == 0)